File: src/talk_diffy_to_me/diff.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Literal

LineKind = Literal["+", "-", " "]

HUNK_RE = re.compile(r"^@@ -(\d+)(?:,(\d+))? \+(\d+)(?:,(\d+))? @@(.*)$")


@dataclass
class DiffLine:
    kind: LineKind
    text: str
    old_ln: int | None
    new_ln: int | None


@dataclass
class Hunk:
    header: str
    old_start: int
    old_count: int
    new_start: int
    new_count: int
    lines: list[DiffLine] = field(default_factory=list)


@dataclass
class File:
    path: str
    old_path: str | None
    status: str  # modified | added | deleted | renamed
    hunks: list[Hunk] = field(default_factory=list)
    binary: bool = False


def parse(text: str) -> list[File]:
    files: list[File] = []
    current: File | None = None
    hunk: Hunk | None = None
    old_ln = 0
    new_ln = 0

    for raw in text.splitlines():
        if raw.startswith("diff --git "):
            parts = raw.split(" ")
            a = parts[2][2:] if parts[2].startswith("a/") else parts[2]
            b = parts[3][2:] if parts[3].startswith("b/") else parts[3]
            current = File(
                path=b,
                old_path=a if a != b else None,
                status="modified",
            )
            files.append(current)
            hunk = None
            continue

        if current is None:
            continue

        if raw.startswith("new file mode"):
            current.status = "added"
        elif raw.startswith("deleted file mode"):
            current.status = "deleted"
            if current.old_path:
                current.path = current.old_path
        elif raw.startswith("rename from "):
            current.status = "renamed"
            current.old_path = raw[len("rename from "):]
        elif raw.startswith("rename to "):
            current.path = raw[len("rename to "):]
        elif raw.startswith("Binary files"):
            current.binary = True
        elif raw.startswith("--- ") or raw.startswith("+++ ") or raw.startswith("index "):
            continue
        elif raw.startswith("@@"):
            m = HUNK_RE.match(raw)
            if not m:
                continue
            os_, oc_, ns_, nc_, _suffix = m.groups()
            hunk = Hunk(
                header=raw,
                old_start=int(os_),
                old_count=int(oc_) if oc_ is not None else 1,
                new_start=int(ns_),
                new_count=int(nc_) if nc_ is not None else 1,
            )
            current.hunks.append(hunk)
            old_ln = hunk.old_start
            new_ln = hunk.new_start
        elif hunk is not None:
            if raw.startswith("\\"):
                continue  # "\ No newline at end of file"
            if raw.startswith("+"):
                hunk.lines.append(DiffLine("+", raw[1:], old_ln=None, new_ln=new_ln))
                new_ln += 1
            elif raw.startswith("-"):
                hunk.lines.append(DiffLine("-", raw[1:], old_ln=old_ln, new_ln=None))
                old_ln += 1
            elif raw.startswith(" "):
                hunk.lines.append(DiffLine(" ", raw[1:], old_ln=old_ln, new_ln=new_ln))
                old_ln += 1
                new_ln += 1
            elif raw == "":
                # empty context line (git sometimes strips the leading space)
                hunk.lines.append(DiffLine(" ", "", old_ln=old_ln, new_ln=new_ln))
                old_ln += 1
                new_ln += 1

    return files

File: internal/diff/diff.go
package diff

import (
	"bufio"
	"fmt"
	"io"
	"strconv"
	"strings"
)

type FileStatus int

const (
	StatusModified FileStatus = iota
	StatusAdded
	StatusDeleted
	StatusRenamed
	StatusCopied
)

func (s FileStatus) String() string {
	switch s {
	case StatusAdded:
		return "added"
	case StatusDeleted:
		return "deleted"
	case StatusRenamed:
		return "renamed"
	case StatusCopied:
		return "copied"
	default:
		return "modified"
	}
}

type Line struct {
	Kind    byte
	Text    string
	OldLine int
	NewLine int
}

type Hunk struct {
	OldStart, OldCount int
	NewStart, NewCount int
	Section            string
	Lines              []Line
}

type File struct {
	OldPath string
	NewPath string
	Status  FileStatus
	Hunks   []Hunk
}

func (f File) DisplayPath() string {
	if f.NewPath != "" && f.NewPath != "/dev/null" {
		return f.NewPath
	}
	return f.OldPath
}

type Diff struct {
	Files []File
}

func Parse(r io.Reader) (*Diff, error) {
	s := bufio.NewScanner(r)
	s.Buffer(make([]byte, 64*1024), 4*1024*1024)

	d := &Diff{}
	var cur *File
	var hunk *Hunk
	var oldLine, newLine int

	finishHunk := func() {
		if hunk != nil && cur != nil {
			cur.Hunks = append(cur.Hunks, *hunk)
			hunk = nil
		}
	}
	finishFile := func() {
		finishHunk()
		if cur != nil {
			d.Files = append(d.Files, *cur)
			cur = nil
		}
	}

	for s.Scan() {
		line := s.Text()

		if strings.HasPrefix(line, "diff --git ") {
			finishFile()
			cur = &File{Status: StatusModified}
			rest := strings.TrimPrefix(line, "diff --git ")
			old, new, ok := splitDiffPaths(rest)
			if !ok {
				return nil, fmt.Errorf("bad diff header: %q", line)
			}
			cur.OldPath, cur.NewPath = old, new
			continue
		}

		if cur == nil {
			if strings.TrimSpace(line) == "" {
				continue
			}
			return nil, fmt.Errorf("unexpected line before diff header: %q", line)
		}

		// file-level metadata
		switch {
		case strings.HasPrefix(line, "new file mode"):
			cur.Status = StatusAdded
			continue
		case strings.HasPrefix(line, "deleted file mode"):
			cur.Status = StatusDeleted
			continue
		case strings.HasPrefix(line, "rename from "):
			cur.Status = StatusRenamed
			cur.OldPath = strings.TrimPrefix(line, "rename from ")
			continue
		case strings.HasPrefix(line, "rename to "):
			cur.NewPath = strings.TrimPrefix(line, "rename to ")
			continue
		case strings.HasPrefix(line, "copy from "):
			cur.Status = StatusCopied
			cur.OldPath = strings.TrimPrefix(line, "copy from ")
			continue
		case strings.HasPrefix(line, "copy to "):
			cur.NewPath = strings.TrimPrefix(line, "copy to ")
			continue
		case strings.HasPrefix(line, "index "),
			strings.HasPrefix(line, "old mode "),
			strings.HasPrefix(line, "new mode "),
			strings.HasPrefix(line, "similarity index "),
			strings.HasPrefix(line, "dissimilarity index "):
			continue
		case strings.HasPrefix(line, "--- "), strings.HasPrefix(line, "+++ "):
			continue
		case strings.HasPrefix(line, "Binary files "):
			continue
		case strings.HasPrefix(line, "@@"):
			finishHunk()
			h, err := parseHunkHeader(line)
			if err != nil {
				return nil, err
			}
			hunk = h
			oldLine = h.OldStart
			newLine = h.NewStart
			continue
		}

		if hunk == nil {
			return nil, fmt.Errorf("unexpected line outside hunk: %q", line)
		}

		if len(line) == 0 {
			hunk.Lines = append(hunk.Lines, Line{Kind: ' ', Text: "", OldLine: oldLine, NewLine: newLine})
			oldLine++
			newLine++
			continue
		}

		kind := line[0]
		text := line[1:]
		switch kind {
		case ' ':
			hunk.Lines = append(hunk.Lines, Line{Kind: ' ', Text: text, OldLine: oldLine, NewLine: newLine})
			oldLine++
			newLine++
		case '+':
			hunk.Lines = append(hunk.Lines, Line{Kind: '+', Text: text, NewLine: newLine})
			newLine++
		case '-':
			hunk.Lines = append(hunk.Lines, Line{Kind: '-', Text: text, OldLine: oldLine})
			oldLine++
		case '\\':
			continue
		default:
			return nil, fmt.Errorf("unexpected line in hunk: %q", line)
		}
	}
	if err := s.Err(); err != nil {
		return nil, err
	}
	finishFile()
	return d, nil
}

func splitDiffPaths(rest string) (string, string, bool) {
	idx := strings.LastIndex(rest, " b/")
	if idx < 0 {
		return "", "", false
	}
	left := rest[:idx]
	right := rest[idx+1:]
	if !strings.HasPrefix(left, "a/") || !strings.HasPrefix(right, "b/") {
		return "", "", false
	}
	return left[2:], right[2:], true
}

func parseHunkHeader(line string) (*Hunk, error) {
	rest := strings.TrimPrefix(line, "@@")
	endIdx := strings.Index(rest, "@@")
	if endIdx < 0 {
		return nil, fmt.Errorf("bad hunk header: %q", line)
	}
	ranges := strings.TrimSpace(rest[:endIdx])
	section := strings.TrimSpace(rest[endIdx+2:])
	parts := strings.Fields(ranges)
	if len(parts) != 2 || !strings.HasPrefix(parts[0], "-") || !strings.HasPrefix(parts[1], "+") {
		return nil, fmt.Errorf("bad hunk header: %q", line)
	}
	oldStart, oldCount, err := parseRange(parts[0][1:])
	if err != nil {
		return nil, fmt.Errorf("bad hunk header %q: %w", line, err)
	}
	newStart, newCount, err := parseRange(parts[1][1:])
	if err != nil {
		return nil, fmt.Errorf("bad hunk header %q: %w", line, err)
	}
	return &Hunk{
		OldStart: oldStart, OldCount: oldCount,
		NewStart: newStart, NewCount: newCount,
		Section: section,
	}, nil
}

func parseRange(s string) (int, int, error) {
	if i := strings.Index(s, ","); i >= 0 {
		start, err := strconv.Atoi(s[:i])
		if err != nil {
			return 0, 0, err
		}
		count, err := strconv.Atoi(s[i+1:])
		if err != nil {
			return 0, 0, err
		}
		return start, count, nil
	}
	start, err := strconv.Atoi(s)
	if err != nil {
		return 0, 0, err
	}
	return start, 1, nil
}
